onnxruntime/contrib_ops/cuda/collective/nccl_kernels.cc

```diff
@@ -107,26 +108,79 @@ Status AllReduce::ComputeInternal(OpKernelContext* context) const {
 
 AllGather::AllGather(const OpKernelInfo& info) : NcclKernel(info) {
   info.GetAttrOrDefault("group_size", &group_size_, static_cast<int64_t>(1));
+  info.GetAttrOrDefault("axis", &axis_, static_cast<int64_t>(0));
+  cuda_ep_ = static_cast<const CUDAExecutionProvider*>(info.GetExecutionProvider());
 }
 
 Status AllGather::ComputeInternal(OpKernelContext* context) const {
   ncclComm_t comm = nccl_->Comm();
 
   auto input_tensor = context->Input<Tensor>(0);
   const void* input_data = input_tensor->DataRaw();
-  const auto in_shape = input_tensor->Shape();
+  const auto& in_shape = input_tensor->Shape();
   int64_t input_count = in_shape.Size();
-  // construct output shape
-  TensorShape out_shape(in_shape);
-  out_shape[0] = group_size_ * out_shape[0];
 
-  void* output_data = context->Output(0, out_shape)->MutableDataRaw();
+  if (axis_ > 0) {
+    // Need transpose
+    std::vector<size_t> permutation(in_shape.NumDimensions());
+    AllocatorPtr alloc;
+    auto status = context->GetTempSpaceAllocator(&alloc);
+    if (!status.IsOK())
+      return status;
+
+    std::iota(std::begin(permutation), std::end(permutation), 0);
+
+    // swap rank 0 and rank axis
+    permutation[axis_] = 0;
+    permutation[0] = axis_;
+    std::vector<int64_t> transposed_input_dims;
+    transposed_input_dims.reserve(in_shape.NumDimensions());
+    for (auto e : permutation) {
+      transposed_input_dims.push_back(in_shape[e]);
+    }
+
+    // Allocate a temporary tensor to hold transposed input
+    auto temp_input = Tensor::Create(input_tensor->DataType(), TensorShape(transposed_input_dims), alloc);
+
+    // Perform the transpose
+    ORT_RETURN_IF_ERROR(onnxruntime::cuda::Transpose::DoTranspose(cuda_ep_->GetDeviceProp(),
```

Comment:
This API looks great. I guess `onnxruntime::cuda::Transpose::DoTranspose` is mainly for eager mode?

Reply:
not really, i guess it is created because we need to perform transpose in some other fused kernels. (although i don't see any gain from this kind of fusion)